File: 19-21.java
package test2;

import java.util.Scanner;

public class TestClass2
{
	public static void main(String[] args)
	{
		@SuppressWarnings("resource")
		Scanner sc = new Scanner(System.in);
		System.out.println("¬ведите цель: "); int w = sc.nextInt();
		System.out.println("¬ведите количество камней в первой куче: "); int i = sc.nextInt();
		System.out.println("¬ведите х: "); int x = sc.nextInt();
		System.out.println("¬ведите у: "); int y = sc.nextInt();
		
		String[][] m = new String[w][w];
		
		for(int i1 = 1; i1 < w; i1++)
		{
			for(int j1 = 1; j1 < w; j1++)
			{
				m[i1][j1] = -1;
				System.out.print(m[i1][j1]);
			}
			System.out.println();
		}
		
		System.out.println();
		System.out.println();
		System.out.print("  ");
		
		for(int i2 = w - 1; i2 > 0; i2--)
			System.out.print(i2 > 9 ? " " + i2 : "  " + i2);

		System.out.println();
		
		for(int i2 = w - 1; i2 > 0; i2--)
		{
			System.out.print(i2 > 9 ? i2 : i2 + " ");
			
			for(int j2 = w - 1; j2 > 0; j2--)
			{
				int i1 = Math.min(i2, j2);
				int j1 = Math.max(i2, j2);

				if (j1 * y + i1 >= w) // < ??????????????????????????????????????????
				{
					m[i1][j1] = 1;
				}
				else if (m[y * i1][j1].charAt(0) == 49 && m[i1][j1 * y].charAt(0) == 49 && m[i1 + x][j1].charAt(0) == 49 && m[i1][j1 + 2].charAt(0) == 49)
				{
					m[i1][j1] = 2;
				}
				else
				{
					m[i1][j1] = 1;
				}
				
				System.out.print(m[i1][j1] < 0 ? " " + m[i1][j1] : "  " + m[i1][j1]);
			}
			System.out.println();
		}
	}
}

File: 19-21.cs
using System;

class Program
{
    static void Main()
    {
        Console.WriteLine("¬ведите конечную цель: ");
        int w = Convert.ToInt32(Console.ReadLine());
        Console.WriteLine("¬ведите кол-во камней в первой куче: ");
        int m = Convert.ToInt32(Console.ReadLine());
        Console.WriteLine("¬ведите ƒействие с сложением: ");
        int x = Convert.ToInt32(Console.ReadLine());
        Console.WriteLine("¬ведите ƒействие с умножением: ");
        int y = Convert.ToInt32(Console.ReadLine());
        
        int[][] a = new int[w * w][w * w]();
        int[][] b = new int[w * w][w * w]();
        int[][] c = new int[w * w][w * w]();
        
        for(int i = w - 1; i > m - 1; i--)
            for(int j = w - 1; i > m - 1; i--)
            {
                if (a[i + x][j] * a[i * y][j] * a[i][j * y] * a[i][j + x] == 0)
                    a[i][j] = 1;
                if (Math.Max(i, j) * y + Math.Min(i, j) >= w)
                {
                    a[i][j] = 1;
                    b[i][j] = 1;
                    c[i][j] = 1;
                }
                else if (a[i][j] == 0)
                {
                    b[i][j] = Math.Min(b[i + x][j], b[i * y][j], b[i][j + x], b[i][j * y]) + 1;
                    c[i][j] = Math.Max(c[i + x][j], c[i * y][j], c[i][j + x], c[i][j * y]) + 1;
                }
                else
                {
                    b[i][j] = Math.Min(b[i + x][j] + (a[i + x][j] == 0 ? 0 : 10), b[i * y][j] + (a[i * y][j] == 0 ? 0 : 10), b[i][j + x] + (a[i][j + x] == 0 ? 0 : 10), b[i][j * y] + (a[i][j * y] == 0 ? 0 : 10)) + 1;
                    c[i][j] = Math.Min(c[i + x][j] + (a[i + x][j] == 0 ? 0 : 10), c[i * y][j] + (a[i * y][j] == 0 ? 0 : 10), c[i][j + x] + (a[i][j + x] == 0 ? 0 : 10), c[i][j * y] + (a[i][j * y] == 0 ? 0 : 10)) + 1;
                }
            }
        
        Console.WriteLine("ќтветы на 19-й номер:");
        Console.WriteLine(Math.Ceiling((w - n) / (y ** 2)));
        Console.WriteLine("ќтветы на 20-й номер:");
        for(int i = 0; i < w; i++)
            if (a[m][i] == 1 && b[m][n] == 3 && c[m][n] == 3) Console.WriteLine(i);
        
        Console.WriteLine("ќтветы на 21-й номер:");
        for(int i = 0; i < w; i++)
            if (a[m][i] == 0 && b[m][n] == 2 && c[m][n] == 4) Console.WriteLine(i);
    }
}